### DISCUSSION TITLE
Support System.Drawing.Image in WPF

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/wpf#5583] [General]
WPF can use **System.Windows.Media.ImageSource** class to provide the image source to WPF controls.
But ImageSource cannot use **System.Drawing.Image** class as a source directly.

System.Windows.Media.ImageSource:
https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.imagesource

If this is a **mixed programming** of Winforms and WPF, the Winforms part needs to use System.Drawing.Image class to display image or bitmap. but now the WPF cannot accept the System.Drawing.Image class.

System.Drawing.Image
https://docs.microsoft.com/zh-cn/dotnet/api/system.drawing.image

**Comment by lindexi** [comment]
@roland5572 You can use [Imaging.CreateBitmapSourceFromHBitmap(IntPtr, IntPtr, Int32Rect, BitmapSizeOptions) Method (System.Windows.Interop)](https://docs.microsoft.com/en-us/dotnet/api/system.windows.interop.imaging.createbitmapsourcefromhbitmap?view=windowsdesktop-5.0) to use the System.Drawing.Image

**Comment by ghost** [comment]
@lindexi 
This requires additional data conversion and may lose performance.
And System.Drawing.Image supports .jpg .png .gif .ico and other image formats, these scenarios may not apply to Bitmap (HBITMAP).

https://docs.microsoft.com/en-us/dotnet/api/system.drawing.imaging.imageformat

If WPF can directly display the Image without additional image saving or rebuilding operation, this can significantly improve performance, especially for large images. image size 1920x1080 or above.

**Comment by lindexi** [comment]
@roland5572  Thank you. And it will call [IWICImagingFactory::CreateBitmapFromHBITMAP (wincodec.h)](https://docs.microsoft.com/en-us/windows/win32/api/wincodec/nf-wincodec-iwicimagingfactory-createbitmapfromhbitmap). 

And WPF will use the WICBitmap type image to render in DirectX

**Comment by ghost** [comment]
@lindexi  It is HBITMAP handle, can it process the .png or .jpg format?

System.Drawing.Image class supports .png and .jpg format.

If it can process the .jpg or .png via HBITMAP handle. I think if .net can provide a public method (eg. ImageSource.LoadImage(System.Drawing.Image image), These operations will be greatly simplified.

stackoverflow has a complex method to convert System.Drawing.Image to WPF ImageSource.

https://stackoverflow.com/questions/41998142/converting-system-drawing-image-to-system-windows-media-imagesource-with-no-resu

**Comment by lindexi** [comment]
@roland5572 Do you means how to show PNG or JPG image in WPF ?

**Comment by madewokherd** [comment]
One could probably make a subclass of `BitmapSource` that implements `BitmapSource.CopyPixels` based on `Bitmap.LockBits`. That should get the best performance possible, as you'd be reading directly from the `Bitmap`'s memory buffer.

**Comment by ghost** [comment]
> @roland5572 Do you means how to show PNG or JPG image in WPF ?

No, Mixed programming, namely WinForms+WPF, is a highlight of .net programing. 

System.Drawing.Image is the main class used to display pictures in Winforms. I think it is very necessary to **directly** support the System.Drawing.Image class in WPF.

If format conversion is required (eg: convert the bytes), this may seriously affect the performance of the program.

In fact, the user does not need to know whether he uses .jpg or .png, he only needs to pass a System.Drawing.Image parameter, and then the program can work normally.

**Comment by Symbai** [comment]
> I think it is very necessary to **directly** support the System.Drawing.Image class in WPF.

The only use-case where this matters is when dealing with third party libraries which only accept System.Drawing.Image. While this feature is good-to-have, it's not _very necessary_. Especially since there are good workarounds (and these workarounds would also be used as an official WPF implementation).

**Comment by ghost** [comment]
> > I think it is very necessary to **directly** support the System.Drawing.Image class in WPF.
> 
> The only use-case where this matters is when dealing with third party libraries which only accept System.Drawing.Image. While this feature is good-to-have, it's not _very necessary_. Especially since there are good workarounds (and these workarounds would also be used as an official WPF implementation).

This feature can simplify the mixed programming of WinForms and WPF. If users or third party libraries don't need mixed programming, it's not very necessary. 

If this feature is available, it will increase the connectivity of program and third-party libraries on nuget and github.

**Comment by singhashish-wpf** [comment]
@roland5572  The issue is moved to discussion as of now.
